### DISCUSSION TITLE
SBERT 추가 실험 결과

### DISCUSSION THREAD

**Discussion by intrandom5** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#75] [Show and tell]
### 초기 실험이 잘 안됐던 이유
https://wikidocs.net/156176
초기에는 위 링크의 글만 보고서 내가 임의로 코드를 구성했었다. 왜냐하면 쉬워 보여서 굳이 다른 코드를 참고할 필요성을 못 느꼈었다.
>
    class SBERT_base_Model(nn.Module):
        def __init__(self, model_name):
            super(SBERT_base_Model, self).__init__():
                self.bert = AutoModel.from_pretrained(model_name)
                self.similarity = nn.CosineSimilarity()
		
        def forward(self, src_ids, tgt_ids):
            u = self.bert(src_ids).last_hidden_state
            v = self.bert(tgt_ids).last_hidden_state
            u_pooling = torch.mean(u, dim=1)
            v_pooling = torch.mean(v, dim=1)
		
            outputs = self.similarity(u_pooling, v_pooling)

            return outputs
그러나 나중에 확인한 결과, 이렇게 학습할 경우 학습을 해도 모델의 파라미터가 갱신이 되지 않는 문제가 있었다..!
### 해결
올바른 코드 구현을 위해 구글링을 조금 더 해봤다.
https://stackoverflow.com/questions/65083581/how-to-compute-mean-max-of-huggingface-transformers-bert-token-embeddings-with-a
여기에서 말하듯, mean pooling과 max pooling을 수행할 때 pad_token 위치의 값들은 빼야한다는 것이다.
지금 보면 너무 당연한데 구현할 땐 상상도 못했다. 당연히 pad token 위치의 값들은 모델 추론에 사용되지 않는 일종의 junk 값인데, 이 값들이 평균 계산에 포함되면 당연히 문제가 있을 것이다. 그렇기 때문에 아래 코드와 같이 attention mask를 이용해 pad_token 위치의 값들은 가리고 평균을 계산해야 한다.
>
    def mean_pooling(model_output, attention_mask):
        token_embeddings = model_output[0]
        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
        return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

코드는 기존 코드를 수정하는게 귀찮아서(대회도 끝난 마당에...) jupyter notebook으로 작성해서 다시 학습을 해봤다.
### 결과
![image](https://user-images.githubusercontent.com/68216991/201243183-8afedc75-3e23-4320-b59a-c9b09e637034.png)
SBERT의 학습 그래프.
위 그래프와 같이 train set에서 loss도 낮아지기 시작하고, dev셋에 대해서도 loss가 많이 낮아진 것을 확인할 수 있다.
![image](https://user-images.githubusercontent.com/68216991/201243340-4ff5d059-513b-4615-9ee2-187add2d27b7.png)
학습 전과 학습 후의 cosine similarity 분포도 확인해 보았다. 위의 그래프는 SBERT 학습 전/후 모델이 dev 셋의 문장 쌍들의 cosine similarity를 측정한 결과의 분포를 나타낸다.
보면 SBERT 학습 전 모델은 앞서 discussion에서 봤듯이, 모든 문장 쌍들의 feature가 높은 cosine similarity를 갖는 것을 볼 수 있다. 그러나 SBERT 학습 후에는 모델이 cosine similarity를 훨씬 넓게 분포시키고 있다는 것을 확인할 수 있다.
이를 봤을 때, SBERT를 통해 임베딩 공간이 넓어졌다라고 생각해도 좋을 것 같다.
### STS 실험
또, 논문을 통해 위의 방식, SBERT는 STS task를 통한 pre-training 방식이라는 것을 발견했다. 그렇기 때문에 위 모델에서 그치지 않고, 2 문장을 [SEP] 토큰으로 이어서 하나의 BERT 모델을 통해 0~5점 사이의 결과를 내는 기존 방식의 STS 모델에 전이학습을 시켜볼 것이다.
SBERT로 학습된 모델을 이용해서 STS training을 수행해보고, SBERT로 학습하지 않은 모델로 STS 학습한 모델과 성능을 비교해 봤다.
![image](https://user-images.githubusercontent.com/68216991/201244528-2affdbb0-e4d7-4dab-8d61-83f82a2cd6fe.png)
결과는 위 그래프와 같았고, SBERT로 학습했던 모델의 성능(갈색)이 더 좋게 나왔다!
최종 score는 다음과 같이 나왔다.
pearson : 0.8905
spearman : 0.8934
두 모델 모두 bert/base를 기반으로 학습되었기 때문에 roberta-large 모델보다는 당연히 성능이 떨어진다.
### 결론
이를 통해 SBERT pretraining 방식도 MLM과 같이 down-stream task에 충분히 좋은 영향을 미친다는 것을 확인할 수 있었다!

**Comment by intrandom5** [comment]
[SBERT.zip](https://github.com/boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13/files/9985823/SBERT.zip)
SBERT 학습에 사용한 SBERT.ipynb 파일은 댓글에 첨부하겠습니다. (압축한 이유는 .ipynb 파일은 올라가지 않아서...)
